### DISCUSSION TITLE
Conversion from Mp3 to Wav offsets the output audio

### DISCUSSION THREAD

**Discussion by WTTSoftwareSolutions** [naudio/NAudio#904] [Q&A] [answered]
When converting an Mp3 to Wav using NAudio 2.10 (and previous versions) the wav file output is longer in length and there seems to be padding added at the beginning, which offsets the actual start of the audio.
```
using NAudio.Wave;

var outFile = @"C:\Temp\NAudioMP3Test\NAudioTestWAV.wav";
var inFile = @"C:\Temp\NAudioMP3Test\NAudioTestMP3.mp3";

using (var reader = new AudioFileReader(inFile)) {
    try {
        WaveFileWriter.CreateWaveFile(outFile, reader);
    } catch (Exception ex) {
        Console.WriteLine(ex.Message);
    }
}
```
![NAudioMp3ToWavTest](https://user-images.githubusercontent.com/58865918/168856082-9fdd776d-c662-4d8d-a051-4d29e5392864.png)

**Comment by markheath** [comment, marked as answer]
This may be inevitable with MP3 - it encodes blocks of audio so the output will not be the exact same length. I expect you'll get a similar result with other tools.

**Reply by WTTSoftwareSolutions** [reply]
I can convert mp3 to wav in Audacity and it is the same length.
I have tried most of the .NET C# solutions including NLayer and MP3Sharp and reading the mp3 always adds the padding.
As far as encoding, NAudio.Lame works perfect. The output mp3 is a match to the input audio.
This would normally not be an issue except we are not generating the mp3 output for listening enjoyment, we are striping one channel of the audio with SMPTE or FSK timecode. The offset is a huge issue with the timecode and the choreography that was done to the original audio.
At this point, I'm at a complete loss. Might have to disable support for mp3.

**Reply by markheath** [reply]
Can you try with both `MediaFoundationReader` and `Mp3FileReader` instead of `AudioFileReader` - do they produce different results

**Reply by WTTSoftwareSolutions** [reply]
I tried both and got the same results. I initially thought that using ```MediaFoundationReader``` would solve the issue, but it gave me the same issue as ```Mp3FileReader```.
I had briefly looked at what Audacity is using, [libmad](https://github.com/markjeee/libmad), but trying to port that to C# was outside the scope of my skills (very rudimentary understanding of C/C++)

**Reply by markheath** [reply]
OK, I suspect that means that the MP3 decoders included with Windows are introducing the delay. 
Two other suggestions for you - (1) you can try plugging NLayer into Mp3FileReader as the decoder. (2) just make a command-line call to FFMPEG.exe if that's possible - I tend to use that a lot these days for converting between various formats

**Reply by WTTSoftwareSolutions** [reply]
To solve the mp3 decoding issue, I've created an ```FFMpegMp3Reader``` which makes use of [FFmpeg.NET](https://github.com/cmxl/FFmpeg.NET). Internally this creates a temp PCM wav file on disk and then reads the temp file in with ```WaveFileReader```.
For the encoding issues, I use [NAudio.Lame](https://github.com/Corey-M/NAudio.Lame) to write the mp3 output.
